Source/Dafny/DafnyAst.cs

```diff
@@ -4698,35 +4732,47 @@ public enum ES { NotYetComputed, Never, ConsultTypeArguments }
 
   public class TupleTypeDecl : IndDatatypeDecl
   {
-    public readonly int Dims;
+    public readonly List<bool> ArgumentGhostness;
+
+    public int Dims
+    {
+      get { return ArgumentGhostness.Count; }
+    }
+    
+    public int NonGhostDims
+    {
+      get { return ArgumentGhostness.Count(x => !x); }
+    }
+
     /// <summary>
     /// Construct a resolved built-in tuple type with "dim" arguments.  "systemModule" is expected to be the _System module.
     /// </summary>
-    public TupleTypeDecl(int dims, ModuleDefinition systemModule, Attributes attributes)
-      : this(systemModule, CreateCovariantTypeParameters(dims), attributes) {
-      Contract.Requires(0 <= dims);
+    public TupleTypeDecl(List<bool> argumentGhostness, ModuleDefinition systemModule, Attributes attributes)
+      : this(systemModule, CreateCovariantTypeParameters(argumentGhostness.Count), argumentGhostness, attributes) {
+      Contract.Requires(0 <= Dims);
```

Comment:
Where does this `Dims` come from? It doesn't seem to make sense as a precondition from the caller's perspective. Perhaps you can move this `Contract.Requires` down a couple of lines and instead state it as a `Contract.Assert`?

Reply:
Uh, good catch. The `Dims` method cannot work yet at that point. It should be `argumentGhostness.Count`.